File: backend/src/main/java/com/project/controller/AuthController.java
package com.project.controller;

import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.Authentication;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import com.project.dto.auth.request.LoginDTO;
import com.project.dto.auth.request.OtpVerifyDTO;
import com.project.dto.auth.request.RegisterDTO;
import com.project.dto.auth.request.RegisterProfessionalDTO;
import com.project.dto.auth.response.LoginResponseDTO;
import com.project.dto.auth.response.RegisterResponseDTO;
import com.project.entities.Professional;
import com.project.entities.User;
import com.project.exception.ApiResponse;
import com.project.exception.ResponseBuilder;
import com.project.security.CustomUserDetails;
import com.project.security.JWTService;
import com.project.service.OtpService;
import com.project.service.UserService;

import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;

@RestController
@RequestMapping("/api/auth")
@RequiredArgsConstructor
public class AuthController {

    private final UserService userService;
    private final AuthenticationManager authenticationManager;
    private final OtpService otpService;
    private final JWTService jwtService;

    /* ================= REGISTER USER ================= */

    @PostMapping("/register")
    public ResponseEntity<ApiResponse<RegisterResponseDTO>> register(
            @RequestBody @Valid RegisterDTO registerDTO) {

        User user = userService.registerUser(registerDTO);

        RegisterResponseDTO response =
                new RegisterResponseDTO(
                        user.getUserId(),
                        "User registered successfully. Please verify OTP."
                );

        return ResponseBuilder.success(
                "User registered successfully",
                response,
                HttpStatus.CREATED
        );
    }

    /* ================= REGISTER PROFESSIONAL ================= */

    @PostMapping("/register-professional")
    public ResponseEntity<ApiResponse<RegisterResponseDTO>> registerProfessional(
            @RequestBody @Valid RegisterProfessionalDTO dto) {

        Professional professional = userService.registerProfessional(dto);

        RegisterResponseDTO response =
                new RegisterResponseDTO(
                        professional.getUser().getUserId(),
                        "Professional registered successfully. Please verify OTP."
                );

        return ResponseBuilder.success(
                "Professional registered successfully",
                response,
                HttpStatus.CREATED
        );
    }

    /* ================= LOGIN ================= */

    @PostMapping("/login")
    public ResponseEntity<ApiResponse<LoginResponseDTO>> login(
            @RequestBody @Valid LoginDTO dto) {

        Authentication authentication =
                authenticationManager.authenticate(
                        new UsernamePasswordAuthenticationToken(
                                dto.getEmail(),
                                dto.getPassword()
                        )
                );

        CustomUserDetails userDetails =
                (CustomUserDetails) authentication.getPrincipal();

        String token = jwtService.generateToken(userDetails);

        return ResponseBuilder.success(
                "Login successful",
                new LoginResponseDTO(token),
                HttpStatus.OK
        );
    }

    /* ================= VERIFY OTP ================= */

    @PostMapping("/verify-otp")
    public ResponseEntity<ApiResponse<Object>> verifyOtp(
            @RequestBody @Valid OtpVerifyDTO dto) {

        otpService.verifyOtp(dto.getUserId(), dto.getOtp());

        return ResponseBuilder.success(
                "OTP verified successfully",
                null,
                HttpStatus.NO_CONTENT
        );
    }
}

File: backend_dotnet/SafeMind.API/Controllers/AuthController.cs
using Microsoft.AspNetCore.Mvc;
using SafeMind.Core.DTOs.Auth;
using SafeMind.Core.DTOs.Auth.Response;
using SafeMind.Core.Interfaces;
using SafeMind.Core.Responses;
using SafeMind.Infrastructure.Data;
using Microsoft.EntityFrameworkCore;
using BCrypt.Net;

namespace SafeMind.API.Controllers
{
    [ApiController]
    [Route("api/auth")]
    public class AuthController : ControllerBase
    {
        private readonly IUserService _userService;
        private readonly IOtpService _otpService;
        private readonly IJwtService _jwtService;
        private readonly ApplicationDbContext _context;

        public AuthController(IUserService userService, IOtpService otpService, IJwtService jwtService, ApplicationDbContext context)
        {
            _userService = userService;
            _otpService = otpService;
            _jwtService = jwtService;
            _context = context;
        }

        [HttpPost("register")]
        public async Task<IActionResult> Register([FromBody] RegisterDTO dto)
        {
            try
            {
                var user = await _userService.RegisterUserAsync(dto);
                var response = new RegisterResponseDTO(user.UserId, "User registered successfully. Please verify OTP.");
                return CreatedAtAction(nameof(Register), new { id = user.UserId }, 
                    ApiResponse<RegisterResponseDTO>.SuccessResponse("User registered successfully. Please verify OTP.", response));
            }
            catch (Exception ex)
            {
                return BadRequest(ApiResponse<object>.FailureResponse(ex.Message));
            }
        }

        [HttpPost("registerProfessional")]
        public async Task<IActionResult> RegisterProfessional([FromBody] RegisterProfessionalDTO dto)
        {
             try
            {
                var professional = await _userService.RegisterProfessionalAsync(dto);
                var response = new RegisterResponseDTO(professional.UserId, "Professional registered successfully. Please verify OTP.");
                return CreatedAtAction(nameof(RegisterProfessional), new { id = professional.ProfessionalId }, 
                    ApiResponse<RegisterResponseDTO>.SuccessResponse("Professional registered successfully. Please verify OTP.", response));
            }
            catch (Exception ex)
            {
                return BadRequest(ApiResponse<object>.FailureResponse(ex.Message));
            }
        }

        [HttpPost("login")]
        public async Task<IActionResult> Login([FromBody] LoginDTO dto)
        {
            var user = await _context.Users.Include(u => u.Role).FirstOrDefaultAsync(u => u.Email == dto.Email);
            if (user == null || !BCrypt.Net.BCrypt.Verify(dto.Password, user.PasswordHash))
            {
                return Unauthorized(ApiResponse<object>.FailureResponse("Invalid email or password"));
            }

            if (!user.IsActive)
            {
                return Unauthorized(ApiResponse<object>.FailureResponse("User is not active. Please verify OTP."));
            }

            var token = _jwtService.GenerateToken(user);
            return Ok(ApiResponse<LoginResponseDTO>.SuccessResponse("Login successful", new LoginResponseDTO(token)));
        }

        [HttpPost("verify-otp")]
        public async Task<IActionResult> VerifyOtp([FromBody] OtpVerifyDTO dto)
        {
            try
            {
                await _otpService.VerifyOtpAsync(dto.UserId, dto.Otp);
                return Ok(ApiResponse<object>.SuccessResponse("OTP verified successfully", null));
            }
            catch (Exception ex)
            {
                return BadRequest(ApiResponse<object>.FailureResponse(ex.Message));
            }
        }
    }
}
